src/IssueViz/Editor/LocationTagging/LocationTagger.cs

```diff
@@ -0,0 +1,188 @@
+
+            // Get the new locations and calculate the spans if necessary
+            var locations = locationService.GetLocations(FilePath);
+            EnsureSpansExist(locations, textSnapshot);
+
+            TagSpans = locations.Select(CreateTagSpan).ToList();
+
+            // Notify the editor that our set of tags has changed
+            var wholeSpan = new SnapshotSpan(textSnapshot, 0, textSnapshot.Length);
+            TagsChanged?.Invoke(this, new SnapshotSpanEventArgs(wholeSpan));
+        }
+
+        private void EnsureSpansExist(IEnumerable<IAnalysisIssueLocationVisualization> locVizs, ITextSnapshot currentSnapshot)
+        {
+            // All spans for issues in the primary file should have been calculated already (whether primary or secondary).
+            // However, there could be secondary locations relating to issues in other files that have not been calculated.
+            foreach (var locViz in locVizs)
+            {
+                if (!locViz.Span.HasValue)
+                {
+                    locViz.Span = spanCalculator.CalculateSpan(locViz.Location, currentSnapshot);
+                }
+            }
+        }
+
+        private static ITagSpan<IIssueLocationTag> CreateTagSpan(IAnalysisIssueLocationVisualization locViz) =>
+            new TagSpan<IIssueLocationTag>(locViz.Span.Value, new IssueLocationTag(locViz));
+
+        private string GetFileName(ITextBuffer textBuffer)
+        {
+            ITextDocument newTextDocument = null;
+            textBuffer.Properties?.TryGetProperty(typeof(ITextDocument), out newTextDocument);
+            return newTextDocument?.FilePath;
+        }
+
+        #region ITagger implementation
+
+        public event EventHandler<SnapshotSpanEventArgs> TagsChanged;
+
+        public IEnumerable<ITagSpan<IIssueLocationTag>> GetTags(NormalizedSnapshotSpanCollection spans)
+        {
+            if (spans.Count == 0 || TagSpans.Count == 0) { yield break; }
+
+            // If the requested snapshot isn't the same as our tags are on, translate our spans to the expected snapshot 
+            if (spans[0].Snapshot != TagSpans[0].Span.Snapshot)
+            {
+                TranslateTagSpans(spans[0].Snapshot);
+            }
+
+            // Find any tags in that overlap with that range
+            foreach (var tagSpan in TagSpans)
+            {
+                if (OverlapsExists(tagSpan.Span, spans))
+                {
+                    yield return tagSpan;
+                }
+            }
+        }
+
+        private void TranslateTagSpans(ITextSnapshot newSnapshot)
+        {
+            var translatedTagSpans = new List<ITagSpan<IIssueLocationTag>>();
+
+            foreach (var old in TagSpans)
+            {
+                var newSpan = old.Span.TranslateTo(newSnapshot, SpanTrackingMode.EdgeExclusive);
+                old.Tag.Location.Span = newSpan; // update the span stored in the location visualization 
```

Comment:
Shouldn't we also call Store.Refresh?
Right now it's been called outside this method, in SafeOnBufferChange, but not on GetTags.
I think it the call to Refresh should be inside this method.

Reply:
I expect that we'd get a buffer changed event in that case before the tags are fetched, so the call to `TranslateTo` in `GetTags` is probably redundant but I'm not sure.

Reply:
We can experiment with removing one of the calls, or we should call Refresh in both cases.